src/types/middleware.ts

```diff
@@ -0,0 +1,31 @@
+import { Dispatch } from './store'
+import { AnyAction } from './actions'
+
+export interface MiddlewareAPI<D extends Dispatch = Dispatch, S = any> {
+  dispatch: D
+  getState(): S
+}
+
+/**
+ * A middleware is a higher-order function that composes a dispatch function
+ * to return a new dispatch function. It often turns async actions into
+ * actions.
+ *
+ * Middleware is composable using function composition. It is useful for
+ * logging actions, performing side effects like routing, or turning an
+ * asynchronous API call into a series of synchronous actions.
+ *
+ * @template DispatchExt Extra Dispatch signature added by this middleware.
+ * @template S The type of the state supported by this middleware.
+ * @template D The type of Dispatch of the store where this middleware is
+ *   installed.
+ */
+export interface Middleware<
+  _DispatchExt = {}, // TODO: remove unused component
```

Comment:
If you guys are going to break versions and go 5.0, now would perhaps be a good time to remove this line?

Reply:
it's actively used in the definition of `applyMiddleware`, so we need to fix that first, I suspect

Reply:
Yep, we're going to get the existing types working with minor tweaks first (this PR) and then make improvements or major breaking changes in subsequent PRs.